yarn-project/pxe/src/private_kernel/hints/compute_tx_include_by_timestamp.ts

```diff
@@ -0,0 +1,57 @@
+import { MAX_INCLUDE_BY_TIMESTAMP_DURATION } from '@aztec/constants';
+import type { PrivateKernelCircuitPublicInputs } from '@aztec/stdlib/kernel';
+import type { UInt64 } from '@aztec/stdlib/types';
+
+const ROUNDED_DURATIONS = [
+  3600, // 1 hour
+  1800, // 30 mins
+  1, // 1 second
+];
+
+function roundTimestamp(blockTimestamp: bigint, includeByTimestamp: bigint, duration: number): UInt64 {
+  const totalDuration = includeByTimestamp - blockTimestamp;
+  const roundedDuration = totalDuration - (totalDuration % BigInt(duration));
+  return blockTimestamp + roundedDuration;
+}
+
+export function computeTxIncludeByTimestamp(
+  previousKernel: PrivateKernelCircuitPublicInputs,
+  maxDuration = MAX_INCLUDE_BY_TIMESTAMP_DURATION,
+): UInt64 {
+  if (maxDuration > MAX_INCLUDE_BY_TIMESTAMP_DURATION) {
+    throw new Error(
+      `Custom max duration cannot be greater than the max allowed. Max allowed: ${MAX_INCLUDE_BY_TIMESTAMP_DURATION}. Custom value: ${maxDuration}.`,
+    );
+  }
+
+  const blockTimestamp = previousKernel.constants.historicalHeader.globalVariables.timestamp;
+  const maxTimestamp = blockTimestamp + BigInt(maxDuration);
+  const includeByTimestamp = previousKernel.includeByTimestamp;
+
+  // If the includeByTimestamp set during the tx execution is greater than or equal to the max allowed duration,
+  // use the maximum allowed timestamp.
+  // Note: It shouldn't be larger than the max allowed duration, but we check for it anyway.
+  if (includeByTimestamp >= maxTimestamp) {
+    return maxTimestamp;
+  }
+
+  // Round it down to the nearest hour/min/second to reduce precision and avoid revealing the exact value.
+  // This makes it harder for others to infer what function calls may have been used to produce a specific timestamp.
+  const roundedTimestamp = ROUNDED_DURATIONS.reduce((timestamp, duration) => {
+    if (timestamp <= blockTimestamp) {
+      // The timestamp must be greater than the block timestamp. Round it down again using a smaller duration.
+      return roundTimestamp(blockTimestamp, includeByTimestamp, duration);
+    }
+    return timestamp;
+  }, 0n);
```

Comment:
I think it would make sense to have this be inside `roundTimestamp` because `roundTimestamp` is just 3 lines and it just feels like it belongs there as it nicely encapsulate the complexity of rounding.